### ISSUE TITLE
[bug]: an exception raised during model installation puts the model hashing thread in a deadlock and prevents installation of the rest of the models

### ISSUE THREAD

**Issue by ljleb** [invoke-ai/InvokeAI#6044]
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### Operating system

Windows

### GPU vendor

Nvidia (CUDA)

### GPU model

RTX 3080

### GPU VRAM

10GB

### Version number

4.0.0rc5

### Browser

Firefox

### What happened

My models directory contains 700+ base/lora/controlnet/etc models/state dicts/files. The initial installation process takes an unreasonable amount of time to complete, but I guess this is a different issue.

Some models in my models directory are not actual stable diffusion models, but instead embeddings. they should really not be "installed" at all. One of my models makes the hashing thread error out during "model installation", like this:

```
[2024-03-24 20:02:26,492]::[ModelInstallService]::INFO --> Model install started: E:/sd/models/Stable-diffusion/[redacted].pt
Exception in thread Thread-1 (_install_next_item):
Traceback (most recent call last):
  File "C:\Python311\Lib\threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "C:\Python311\Lib\threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "D:\Program Files\InvokeAI\.venv\Lib\site-packages\invokeai\app\services\model_install\model_install_default.py", line 458, in _install_next_item
    self._register_or_install(job)
  File "D:\Program Files\InvokeAI\.venv\Lib\site-packages\invokeai\app\services\model_install\model_install_default.py", line 486, in _register_or_install
    key = self.register_path(job.local_path, job.config_in)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Program Files\InvokeAI\.venv\Lib\site-packages\invokeai\app\services\model_install\model_install_default.py", line 176, in register_path
    return self._register(model_path, config)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Program Files\InvokeAI\.venv\Lib\site-packages\invokeai\app\services\model_install\model_install_default.py", line 625, in _register
    info = info or ModelProbe.probe(model_path, config, hash_algo=self._app_config.hashing_algorithm)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Program Files\InvokeAI\.venv\Lib\site-packages\invokeai\backend\model_manager\probe.py", line 140, in probe
    model_type = cls.get_model_type_from_checkpoint(model_path)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Program Files\InvokeAI\.venv\Lib\site-packages\invokeai\backend\model_manager\probe.py", line 223, in get_model_type_from_checkpoint
    if any(key.startswith(v) for v in {"cond_stage_model.", "first_stage_model.", "model.diffusion_model."}):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Program Files\InvokeAI\.venv\Lib\site-packages\invokeai\backend\model_manager\probe.py", line 223, in <genexpr>
    if any(key.startswith(v) for v in {"cond_stage_model.", "first_stage_model.", "model.diffusion_model."}):
           ^^^^^^^^^^^^^^
AttributeError: 'int' object has no attribute 'startswith'
```

Some of these files are situated under torrents or git lfs repos, so moving or deleting them is not a practical solution.

Whenever InvokeAI encounters an embedding in the models directory, it stops installing any other already queued model install. The thread seems to be in some sort of deadlock. I don't know how this generalizes to other types of unexpected/broken models, but it appears to happen with sd1.5 embeddings.

When I try to install the models manually again from the UI after entering this soft lock state, this is printed in the console multiple times, once per not-yet-installed model:

```
[2024-03-24 20:31:53,870]::[ModelInstallService]::WARNING --> There is already an active install job for E:/sd/models/Stable-diffusion/pure/sd_xl_turbo_1.0_fp16.safetensors. Not enqueuing.
```

Additionally, when in this soft lock state, the server cannot close with CTRL+C. It prints this:

```
[2024-03-24 21:16:39,029]::[ModelInstallService]::WARNING --> Cancelling job {job.id}
```

And then nothing happens. No matter how many times I spam CTRL+C again or wait, it just never exits. I have to close the terminal / go in task manager to close it.

A major problem to this is that to "install" existing or future new models with the automatic model discovery feature, I have to constantly remove these unrelated files from my models directory because otherwise I need to completely restart invokeai to fix the soft lock state.

Regardless of why a model fails to load or is invalid, a single error during model installation shouldn't put the server in a buggy state.

### What you expected to happen

I expect InvokeAI to skip model installations that raise an exception instead of entering the strange soft lock state described above.

### How to reproduce the problem

1. Move an embedding to C:\models-dir (or any other models dir of your choice)
2. Start InvokeAI
3. Go to Model Manager > Add Model > Scan Folder
4. Set Folder to C:\models-dir
5. Click "Scan Folder"
6. Click "Install All"
7. Wait for the server to try to install the embedding, which should raise an exception
8. Note how the server halts the installation of any pending model

### Additional context

This seems to only happen when there are invalid/unexpected model types in the models directory.

### Discord username

ljleb

**Comment by psychedelicious** [comment]
The performance issue is probably model hashing. You can opt out of this entirely: https://invoke-ai.github.io/InvokeAI/features/CONFIGURATION/#model-hashing

Some context - we do "install" embeddings. This means we:
- "Probe" the file to determine its attributes (this is what is erroring)
- Store some information about the model in the database to facilitate listing and interacting with it

Can you please upload one of the problematic embeddings? If it is in a folder, please zip and upload the whole folder. We'll need this to figure out why the probing isn't successful.

I've addressed the soft-lock issue in #6048.

**Comment by ljleb** [comment]
> I've addressed the soft-lock issue in https://github.com/invoke-ai/InvokeAI/pull/6048.

Awesome, I appreciate the fast pace on this issue!

> Can you please upload one of the problematic embeddings? If it is in a folder, please zip and upload the whole folder. We'll need this to figure out why the probing isn't successful.

I am not sure whether I can share the file directly for license reasons...

I don't really mind if invoke cannot recognize this type of file, as long as it doesn't break down because of it, then I would personally consider this issue resolved.

If that can be of any help, I see that the structure of the problematic file is this, upon inspection with torch.load:

```
{
    320: (
        {
            "linear1.weight": tensor(shape=[640, 320]),
            "linear1.bias": tensor(shape=[640]),
            "linear2.weight": tensor(shape=[320, 640]),
            "linear2.bias": tensor(shape=[320])
        },
        {
            "linear1.weight": tensor(shape=[640, 320]),
            "linear1.bias": tensor(shape=[640]),
            "linear2.weight": tensor(shape=[320, 640]),
            "linear2.bias": tensor(shape=[320])
        }
    ),
    640: (
        {
            "linear1.weight": tensor(shape=[1280, 640]),
            "linear1.bias": tensor(shape=[1280]),
            "linear2.weight": tensor(shape=[640, 1280]),
            "linear2.bias": tensor(shape=[640])
        },
        {
            "linear1.weight": tensor(shape=[1280, 640]),
            "linear1.bias": tensor(shape=[1280]),
            "linear2.weight": tensor(shape=[640, 1280]),
            "linear2.bias": tensor(shape=[640])
        }
    ),
    768: (
        {
            "linear1.weight": tensor(shape=[1536, 768]),
            "linear1.bias": tensor(shape=[1536]),
            "linear2.weight": tensor(shape=[768, 1536]),
            "linear2.bias": tensor(shape=[768])
        },
        {
            "linear1.weight": tensor(shape=[1536, 768]),
            "linear1.bias": tensor(shape=[1536]),
            "linear2.weight": tensor(shape=[768, 1536]),
            "linear2.bias": tensor(shape=[768])
        }
    ),
    1280: (
        {
            "linear1.weight": tensor(shape=[2560, 1280]),
            "linear1.bias": tensor(shape=[2560]),
            "linear2.weight": tensor(shape=[1280, 2560]),
            "linear2.bias": tensor(shape=[1280])
        },
        {
            "linear1.weight": tensor(shape=[2560, 1280]),
            "linear1.bias": tensor(shape=[2560]),
            "linear2.weight": tensor(shape=[1280, 2560]),
            "linear2.bias": tensor(shape=[1280])
        }
    )
}
```

Let me know if I can provide any other type of information that could be useful to resolve the problem.

**Comment by psychedelicious** [comment]
Thanks, I think that's enough info to understand the issue. This kind of model appears to use integers for keys in its state dict. Our model probing logic assumes the keys are strings, and we try to do string things to the keys - like checking if the key starts with a certain string.

Of course, this doesn't work for integers, and we get this error:
`AttributeError: 'int' object has no attribute 'startswith'`

I've handled this situation in #6051. These models won't work in invoke, because we don't have a good heuristic for determining what they actually are, but at least we get a more sensible error now (e.g. `InvalidModelConfig`). Since you are OK with the models not working, I've marked this issue to be closed when that PR is merged.
